wgpu-core/src/device/queue.rs

```diff
@@ -458,17 +467,133 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
+
+            let mut required_buffer_inits: FastHashMap<
+                id::BufferId,
+                Vec<Range<wgt::BufferAddress>>,
+            > = FastHashMap::default();
+
+            for &cmb_id in command_buffer_ids {
+                let cmdbuf = command_buffer_guard
+                    .get(cmb_id)
+                    .map_err(|_| QueueSubmitError::InvalidCommandBuffer(cmb_id))?;
+
+                for buffer_use in cmdbuf.buffer_memory_init_actions.iter() {
+                    let buffer = buffer_guard
+                        .get_mut(buffer_use.id)
+                        .map_err(|_| QueueSubmitError::DestroyedBuffer(buffer_use.id))?;
+
+                    let uninitialized_ranges = buffer
+                        .initialization_status
+                        .drain_uninitialized_ranges(buffer_use.range.clone());
+                    match buffer_use.kind {
+                        MemoryInitKind::ImplicitlyInitialized => {
+                            uninitialized_ranges.for_each(drop);
+                        }
+                        MemoryInitKind::NeedsInitializedMemory => {
+                            required_buffer_inits
+                                .entry(buffer_use.id)
+                                .or_default()
+                                .extend(uninitialized_ranges);
+                        }
+                    }
+                }
+            }
+            required_buffer_inits
+        };
+
+        // Memory init is expected to be rare (means user relies on default zero!), so most of the time we early here!
+        if required_buffer_inits.is_empty() {
+            return Ok(());
+        }
+
+        let (mut device_guard, mut token) = hub.devices.write(&mut token);
+        let (buffer_guard, _) = hub.buffers.read(&mut token);
+        let device = device_guard
+            .get_mut(queue_id)
+            .map_err(|_| DeviceError::Invalid)?;
+
+        device
+            .pending_writes
+            .dst_buffers
+            .extend(required_buffer_inits.keys());
+        device.borrow_pending_writes(); // Call ensures there is a pending_writes cmdbuffer, but using the reference returned would make the borrow checker unhappy!
+        let pending_writes_cmd_buf = device.pending_writes.command_buffer.as_mut().unwrap();
+        let mut trackers = device.trackers.lock();
+
+        for (buffer_id, mut ranges) in required_buffer_inits.drain() {
+            // Collapse touching ranges. We can't do this any earlier since we only now gathered ranges from several different command buffers!
+            ranges.sort_by(|a, b| a.start.cmp(&b.start));
+            for i in (1..ranges.len()).rev() {
+                assert!(ranges[i - 1].end <= ranges[i].start); // The memory init tracker made sure of this!
+                if ranges[i].start == ranges[i - 1].end {
+                    ranges[i - 1].end = ranges[i].end;
+                    ranges.remove(i);
```

Comment:
uuh, that's going to make it quadratic complexity O(N^2)
Let's find a better way to collapse them?